### DISCUSSION TITLE
Slide transitions for Marp have made stably available now!

### DISCUSSION THREAD

**Discussion by yhatt** [marp-team/marp#168] [Announcements]
In the latest Marp, you can easily add transition effects working on HTML presentation ([`bespoke` template](https://github.com/marp-team/marp-cli#bespoke-template-default)) to your Markdown slides, powered by [View Transitions API](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API). <img src="https://cultofthepartyparrot.com/parrots/hd/shuffleparrot.gif" valign="top" width="16" height="16">

## Prerequisite

The slide transition depends on **[View Transitions API](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API)** in the using browser. To show transition effects, a viewer has to show HTML slide in the browser which have supported View Transitions API.

- **Chrome**: ✅ (111-)
- **Edge**: ✅ (111-)
- **Firefox**: ✅ (144-)
- **Safari**: ✅ (18-)

## Example: `fade` transition

Both of Marp CLI and Marp for VS Code have supported [`transition` local directive](https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md#transition-local-directive).

```markdown
---
marp: true
transition: fade
---

...
```

<p align="center">
  <img src="https://raw.githubusercontent.com/marp-team/marp-cli/main/docs/bespoke-transitions/images/fade.gif" alt="Fade transition effect" width="200" />
</p>

[Marp CLI example](https://github.com/yhatt/marp-cli-example) (by the Marp maintainer: https://yhatt.github.io/marp-cli-example/) has a fade transition in every pages.

### 33 built-in transitions

Marp has built-in **33 transition effects** out of the box, and you can change the kind of transitions with the directive.

[You can see all of these transition effects in action on the documentation of Marp CLI](https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md#built-in-transitions), or [auto-completion help for `transition` local directive](https://github.com/marp-team/marp-vscode#auto-completion) provided by Marp for VS Code.

- **[Marp CLI documentation: Built-in transitions](https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md#built-in-transition)** (**NOTICE**: There are a lot of visual stimulus)

## Define custom transitions

You also can define the custom transitions and animations, both in theme CSS and Markdown inline style. We provide unlimited extensibility of your own transitions with your creativity.

```css
/* Define "triangle" transition effect */
@keyframes marp-incoming-transition-triangle {
  from { clip-path: polygon(0% 0%, 0% 0%, 0% 0%); }
  to { clip-path: polygon(0% 0%, 200% 0%, 0% 200%); }
}
@keyframes marp-incoming-transition-backward-triangle {
  from { clip-path: polygon(100% 100%, 100% 100%, 100% 100%); }
  to { clip-path: polygon(-100% 100%, 100% -100%, 100% 100%); }
}
```

- **[Marp CLI documentation: Custom transitions](https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md#custom-transitions)**
- **[[Blog article] Marp CLI: How to make custom transition](https://marp.app/blog/how-to-make-custom-transition)**

<a href="https://marp.app/blog/how-to-make-custom-transition">
  <p align="center">
    <img src="https://marp.app/og-images/how-to-make-custom-transition.jpg" alt="Marp CLI: How to make custom transition" width="300" />
  </p>
</a>

## Morphing animations

With sprinkling just a little bit of CSS, you can use morphing effects in Marp that are similar to PowerPoint's Morph or Keynote's Magic Move. It should be useful to help the audience understand the flow of things between pages.

It may sound unbelievable, but this is [actually part of the web standard (proposal)!](https://developer.mozilla.org/en-US/docs/Web/CSS/view-transition-name)

- **[Marp CLI documentation: Morphing animations](https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md#morphing-animations)**

<p align="center">
  <img src="https://raw.githubusercontent.com/marp-team/marp-cli/main/docs/bespoke-transitions/images/morphing-animation.gif" alt="Marp CLI morphing animations" width="300" />
</p>

## Reduce transition effects

If the browser has detected that the viewer set a setting an operating system preference to reduce motions and animations in the interface, every effects by transitions will force to a simple `fade` animation.

If you want to know details about why required this, see the article "[`prefers-reduced-motion`: Sometimes less movement is more](https://web.dev/prefers-reduced-motion/)" by [web.dev](https://web.dev/).

## Examples

I'm providing the showcase of built-in transitions, creative examples about custom transitions, and a simple example of morphing animation at ~~[Glitch](https://glitch.com/~marp-cli-page-transitions)~~.

- **~~[Marp CLI page transition showcase](https://marp-cli-page-transitions.glitch.me/)~~**
- **~~[Marp CLI custom transitions example](https://marp-cli-page-transitions.glitch.me/custom.html)~~**
- **~~[Marp CLI transition with morphing animation](https://marp-cli-page-transitions.glitch.me/morph.html)~~**

<a href="https://marp-cli-page-transitions.glitch.me/">
  <p align="center">
    <img src="https://cdn.glitch.global/681769ad-39a3-4267-a2d4-65621076732b/og-image.jpg?v=1676828241038" alt="Marp CLI page transition showcase" width="300" />
  </p>
</a>

> [!NOTE]
>
> [Glitch has been shut down](https://blog.glitch.com/post/changes-are-coming-to-glitch) so examples are currently not available. I'll recover them sequentially as soon as prepared.

## Documentation

To see the documentations about the overall of the transition feature provided by Marp CLI `bespoke` template, check out https://github.com/marp-team/marp-cli/blob/main/docs/bespoke-transitions/README.md.

> Related issue: https://github.com/marp-team/marp-cli/issues/447

**Comment by chenxizhang** [comment]
very cool

**Comment by macbuse** [comment]
Hi 

I just can't get this to work sorry. I did :

>>curl -o ./showcase.md https://gist.githubusercontent.com/yhatt/d9e86ee53eb8816aaf9c996e773b6f82/raw/transition-showcase.md

>>npx @marp-team/marp-cli@^2 --bespoke.transition --preview transition-showcase.md 

but only the initial fade transition seems to work.

>>marp -- version

@marp-team/marp-cli v2.2.2 (w/ user-installed @marp-team/marp-core v2.4.2)

Chrome Version 108.0.5359.124 (Official Build) (64-bit)

---

**Reply by yhatt** [reply]
If Chrome by non-stable channel has been installed, such as beta, dev, and canary, Marp CLI will prefer them than a stable Chrome (108). Don't you have one of them installed?

These versions have no longer supported experimental "Shared Elemental Transition" browser API in favor of the latest spec "View Transition API" (still experimental). So Marp CLI transitions are not working with Chrome 109 and later.

- https://github.com/marp-team/marp-cli/issues/483


[Chrome 109 is going to become stable soon.](https://chromestatus.com/) In next weekend, I will work on the update of Marp CLI to switch API from Shared Elemental Transition API to View Transition API completely.

- https://github.com/marp-team/marp-cli/pull/489

**Reply by yhatt** [reply]
**UPDATE:** If you still have seen fade transitions in every kind of transitions with the stable version, perhaps [you may have set a preference about reduced motion in your OS/device](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion#user_preferences). Please confirm the setting.

**Comment by macbuse** [comment]
Hi

Thanks very much for your reply. I've been using reveal.js for slides and I
want to switch to Marp because of issues with pandoc/mathjax. I look
forward to your next iteration.

On Wed, 4 Jan 2023 at 17:25, Yuki Hattori ***@***.***> wrote:

> If Chrome by non-stable channel has been installed, such as beta, dev, and
> canary, Marp CLI will prefer them than a stable Chrome (108). Don't you
> have one of them installed?
>

No. I have a Chromium Version 108.0.5359.71 (Official Build) Built on
Ubuntu , running on Ubuntu 18.04 (64-bit)


Message ID: ***@***.***
> com>
>


-- 
greg mc

Grenoble France

https://www-fourier.ujf-grenoble.fr/~mcshane/

**Comment by macbuse** [comment]
Dear Hattori-san,

I'm sorry I had forgotten about this.

I installed the latest MARP standalone
and the transitions work on my Chromebook
with Chrome 112.0.5615.134
and Chromium 115.0.5759.0

Thank you for your work,

On Sat, 18 Mar 2023 at 15:43, Yuki Hattori ***@***.***> wrote:

> *UPDATE:* If you still have seen fade transitions in every kind of
> transitions with the stable version, perhaps you may have set a
> preference about reduced motion in your OS/device
> ***@***.***/prefers-reduced-motion#user_preferences>.
> Please confirm the setting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/marp-team/discussions/168#discussioncomment-5353702>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AELF7COMIK7QLE4PBS5PWLDW4XCYHANCNFSM5EIUIYUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
greg mc

Grenoble France

https://www-fourier.ujf-grenoble.fr/~mcshane/

**Comment by xkeshav** [comment]
does this transition feature support in vs code extension? I have tried but in Preview Pane I cant see any transition.  using like below.

vs code: 1.86
Marp for VS Code v 2.8.0
mac OS v 14.2.0

```md
---
marp: true
transition: fade
theme: gaia
paginate: true
backgroundColor: #ffa
footer: '**@xkeshav**'
---

```

**Reply by yhatt** [reply]
The Markdown preview pane in VS Code does not have the concept of "switching pages," so transitions will not be displayed. The actual behavior can be seen in the IntelliSense help and the exported HTML.

**Comment by boldandbusted** [comment]
Just FYI, https://github.com/mozilla/standards-positions/issues/677#event-9514679616 has been marked as completed. Does that mean this works with Firefox now? Thanks.

**Reply by yhatt** [reply]
It means Mozilla's position for View Transitions API has been confirmed as positive, but does not mean already available in Firefox.You can peek [Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1823896) for the progress of implementation for Firefox.

**Comment by kirnhans** [comment]
Does Marp support in-slide animations? Like having text appear on the slide on-click, without changing the slide.

**Reply by yhatt** [reply]
Yes. Marp's HTML slides support slide transitions, simple build lists through [Fragmented list](https://marpit.marp.app/fragmented-list) feature, and autoplay animations [using CSS Animations](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_animations/Using_CSS_animations). Using Fragmented list would be best to use in your use case.

**Reply by kirnhans** [reply]
Thank you for the prompt response! Is there any prospect of this feature being added to the pptx slides? It must be quite a bit more complicated.

**Comment by yhatt** [comment]
#### Transitions are now working on all modern browsers 🎉

I'm happy to announce that transitions are now working in all stable versions of modern browsers with [the release of Firefox 144](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/144#css). [The documentation of Marp CLI](https://github.com/marp-team/marp-cli/tree/main/docs/bespoke-transitions) will also be updated soon.

Enjoy creating more effective slides using transitions 🥳

**Comment by Beurt** [comment]
This is huge! Thank you. I am wondering if transition (fade for instance) are possible on `data-marpit-fragments`?

**Reply by yhatt** [reply]
Transitions between slides and [the fragmented list](https://marpit.marp.app/fragmented-list) are different functionalities. If you want to animate fragments within a slide, use regular [CSS transitions](https://developer.mozilla.org/docs/Web/CSS/Guides/Transitions) instead.

<details>
<summary>Example</summary>

```markdown
<style>
  [data-bespoke-marp-fragment] {
    /* Set transition for fragment opacity */
    transition: opacity 0.5s linear;
  }

  [data-bespoke-marp-fragment="inactive"] {
    /* Revert the default style for inactive fragments by Marp CLI's bespoke template (visibility: hidden) */
    visibility: visible !important;

    /* Disable pointer events and interactivity for inactive fragments */
    pointer-events: none;
    interactivity: inert; /* Only supported in Chrome 135 and later */

    /* Set opacity to 0 */
    opacity: 0;

    @starting-style {
      /* Prevent unexpected fade-out transition on loading slide */
      opacity: 0;
    }
  }
</style>

# Hello

* Fragmented list item 1
* Fragmented list item 2
* Fragmented list item 3
```

</details>

**Reply by Beurt** [reply]
Of course, sorry I mismatched CSS and Slide transitions. Thank you for the example, it is very effective and seems pretty useful. I think it could help many others than me.

**Comment by pke** [comment]
Would parallax transitions be possible too? I don't know how one would define them, but I could imagine multiple backgrounds that are overlayed on top of each other?

**Reply by yhatt** [reply]
Marp's slide transitions are based on animations between single layers, so complex transitions like parallax effects are generally difficult. Check out https://marp.app/blog/how-to-make-custom-transition#the-anatomy-of-a-transition.

However, by using the [`view-transition-name` CSS property](https://developer.mozilla.org/docs/Web/CSS/Reference/Properties/view-transition-name), you can "detach" content within a slide from the transition. This might allow you to achieve a parallax effect for multiple images. ref. https://github.com/marp-team/marp-cli/tree/main/docs/bespoke-transitions#morphing-animations

<details>
<summary>Example</summary>

```markdown
---
transition: cover
paginate: true
---

<style>
  /* Detach every contents in the slide from the root transition group */
  section > *, section::before, section::after {
    view-transition-name: match-element;
  }

  /* Background image overlay */
  img[alt="bg2"] {
    view-transition-name: bg2;

    position: absolute;
    display: block;
    width: 240px;
    height: auto;
    box-shadow: 0 4px 12px rgb(0 0 0 / .75);
  }
</style>

# Page 1

Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quidem minima dicta fugit aspernatur. Omnis molestias quam numquam provident! Et repellat eaque quidem. Culpa, distinctio. Error itaque officia velit facere eos.

![bg opacity:0.3](https://picsum.photos/1920?image=29)
![bg2](https://picsum.photos/600/400?image=28)

<style scoped>
  img[alt="bg2"] {
    left: 80px;
    top: 500px;
    rotate: 5deg;
  }
</style>

---

# Page 2

Lorem ipsum dolor sit, amet consectetur adipisicing elit. Tempore deserunt in beatae aut facilis, impedit obcaecati saepe ratione dignissimos accusantium autem laudantium nesciunt ex modi sit? Dolorum non nesciunt a!

![bg opacity:0.3](https://picsum.photos/1920?image=30)
![bg2](https://picsum.photos/600/400?image=32)

<style scoped>
  img[alt="bg2"] {
    right: 80px;
    top: 500px;
    rotate: -5deg;
  }
</style>
```

</details>

Since the detached content will be morphed between the 2 slides, you won't be able to set up detailed animations. However, depending on how you move the detached images on the background, it might produce better suggestive effect. It's up to you how you move the images.

**Reply by pke** [reply]
Wow, thanks @yhatt! I was not expecting such a swift response. I will give it a try!
Thanks!
